e2e/testcases/basic_test.go

```diff
@@ -0,0 +1,365 @@
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Remove dir namespace, add subdir namespace")
+	nt.WaitForRepoSyncs()
+
+	if err := nt.Validate("subdir", "", &corev1.Namespace{}); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	if err := nt.ValidateNotFound("dir", "", &corev1.Namespace{}); err != nil {
+		nt.T.Fatal("Namespace still exist when it should have been converted")
+	}
+}
+
+func TestSyncDeploymentAndReplicaSet(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation2)
+
+	// Test the ability to fix a mistake: overlapping replicaset and deployment.
+	// Readiness behavior is undefined for this race condition.
+	// One or both of the Deployment and ReplicaSet may become unhealthy.
+	// But regardless, the user should be able to correct the situation.
+	nt.T.Log("Add a replicaset")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/dir-namespace.yaml", yamlDir), "acme/namespaces/dir/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/replicaset-helloworld.yaml", yamlDir), "acme/namespaces/dir/replicaset.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add replicaset")
+
+	// This sync may block until reconcile timeout is reached,
+	// because ReplicaSet or Deployment may never reconcile.
+	// So this wait timeout must be longer than the reconcile timeout (5m).
+	nt.WaitForRepoSyncs()
+	nt.T.Log("check that the replicaset was created")
+	if err := nt.Validate("hello-world", "dir", &appsv1.ReplicaSet{}, nomostest.HasLabel("app", "hello-world")); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Add a corresponding deployment")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/deployment-helloworld.yaml", yamlDir), "acme/namespaces/dir/deployment.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add corresponding deployment")
+	nt.WaitForRepoSyncs()
+
+	nt.T.Log("check that the deployment was created")
+	if err := nt.Validate("hello-world", "dir", &appsv1.Deployment{}); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Remove the deployment")
+	nt.RootRepos[configsync.RootSyncName].Remove("acme/namespaces/dir/deployment.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Remove deployment")
+	// This sync may block until reconcile timeout is reached,
+	// because the ReplicaSet is re-applied before deleting the Deployment.
+	// So this wait timeout must be longer than the reconcile timeout (5m).
+	nt.WaitForRepoSyncs()
+
+	nt.T.Log("check that the deployment was removed and replicaset remains")
+	if err := nt.ValidateNotFound("hello-world", "dir", &appsv1.Deployment{}); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := nt.Validate("hello-world", "dir", &appsv1.ReplicaSet{}, nomostest.HasLabel("app", "hello-world")); err != nil {
+		nt.T.Fatal(err)
+	}
+}
+
+// RoleBindingsHasName will check the Rolebindings name and compare it with expected value
+func RoleBindingsHasName(expectedName string) nomostest.Predicate {
```

Comment:
`RoleBindingsHasName ` -> `RoleBindingHasName`
Also this seems like a pretty generic function that can be moved to `nomostest`